### PR TITLE
Docking cacheing

### PR DESCRIPTION
**Pull request by hmacdope**
Alternate implementation of #672, regenerating the docking result in memory from a saved .json.

## Developers certificate of origin
- [X] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

**Comment by hmacdope** [comment]
@apayne97 this is a slightly different version of docking results caching  heavily inspired by and borrowing from (thank you) your work in #672, but using a json of the DockingResults object and delegating cacheing to the docking subclass. Lets discuss tommorow at dev sync.

**Comment by apayne97** [comment]
Yeah I like these changes you've made! Thanks for taking the baton and running with it on this

File: asapdiscovery-docking/asapdiscovery/docking/docking_v2.py [hunk 1]
```diff
@@ -236,3 +227,42 @@ def make_df_from_docking_results(results: list["DockingResult"]):
...
+        output_pref = result.unique_name()
+        compound_dir = output_dir / output_pref
+        compound_dir.mkdir(parents=True, exist_ok=True)
+        output_sdf_file = compound_dir / "docked.sdf"
+        output_pdb_file = compound_dir / "docked_complex.pdb"
+        output_json_file = compound_dir / "docking_result.json"
+        result.posed_ligand.to_sdf(output_sdf_file)
+        combined_oemol = result.to_posed_oemol()
+        save_openeye_pdb(combined_oemol, output_pdb_file)
+        result.to_json_file(output_json_file)
```
Comment (ijpulidos) [inline asapdiscovery-docking/asapdiscovery/docking/docking_v2.py:254]: Not a blocker. Maybe I'm missing something, but is there a reason why we need to write our own `to_json_file` implementation insetad of using [pydantic's one](https://docs.pydantic.dev/1.10/usage/exporting_models/#modeljson)?

**Review by ijpulidos — APPROVED** [review]
It is looking good, I have a few non-blocking comments/suggestions. Thanks!

**Comment by ijpulidos** [comment]
Oh, I just realized that the docking-ci isn't passing. We probably want to solve that first before merging these changes.

**Comment by ijpulidos** [comment]
Possibly related to some log file being already closed and we would need to open it again or something like that? https://github.com/choderalab/asapdiscovery/actions/runs/7123521130/job/19396270676?pr=679#step:8:123

File: asapdiscovery-docking/asapdiscovery/docking/docking_v2.py (hunk 1 above)
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/docking_v2.py:254]: Ah just lightly wraps the `json()` call from pydantic in writing to a file AFAIK the linked function just dumps the json in memory.

File: asapdiscovery-docking/asapdiscovery/docking/openeye.py [hunk 2]
```diff
@@ -47,6 +48,55 @@ class POSIT_RELAX_MODE(Enum):
...
+class POSITDockingResults(DockingResult):
+    """
+    Schema for a DockingResult from OEPosit, containing both a DockingInputPair used as input to the workflow
+    and a Ligand object containing the docked pose.
+    """
+
+    type: Literal["POSITDockingResults"] = "POSITDockingResults"
+
+    @staticmethod
+    def make_df_from_docking_results(results: list["DockingResult"]):
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/openeye.py:60]: Yep good idea.

File: asapdiscovery-docking/asapdiscovery/docking/openeye.py [hunk 3]
```diff
@@ -142,112 +193,135 @@ def _dock(
...
-                    raise ValueError(f"Unknown error handling option {error}")
+                    if error == "skip":
+                        logger.warn(
+                            f"docking failed for input pair with compound name: {set.ligand.compound_name}, smiles: {set.ligand.smiles} and target name: {set.complex.target.target_name}"
+                        )
+                        docking_results.append(None)
+                    elif error == "raise":
+                        raise ValueError(
+                            f"docking failed for input pair with compound name: {set.ligand.compound_name}, smiles: {set.ligand.smiles} and target name: {set.complex.target.target_name}"
+                        )
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/openeye.py]: Done

File: asapdiscovery-docking/asapdiscovery/docking/docking_v2.py [hunk 4]
```diff
@@ -236,3 +227,42 @@ def make_df_from_docking_results(results: list["DockingResult"]):
...
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, DockingResult):
+            raise NotImplementedError
+
+        # Just check that both Complexs and Ligands are the same
+        return (
+            (self.input_pair == other.input_pair)
+            and (self.posed_ligand == other.posed_ligand)
+            and (self.probability == other.probability)
```
Comment (chrisiacovella) [inline asapdiscovery-docking/asapdiscovery/docking/docking_v2.py:264]: Would it make sense to use isclose on the probability comparison?

Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/docking_v2.py:264]: Good idea.

**Comment by hmacdope** [comment]
Thanks @chrisiacovella, much appreciated.

**Pull request merged**